dysnix/bsc/templates/scripts/_init_from_gcs.tpl

```diff
@@ -97,45 +98,77 @@ TIMESTAMP_0="$(${S5CMD} cat s3://${TIMESTAMP_URL})"
 
 # we're ready to perform actual data sync
 
-# we're done when both are true
+# we're done when all are true
 # 1) start and stop timestamps did not changed during data sync - no process started or finished updating the cloud
-# 2) 0 objects copied
+# 2) start timestamp is before stop timestamp - no process is in progress updating the cloud
+# 3) 0 objects copied
 SYNC=2
 CLEANUP=1
 while [ "${SYNC}" -gt 0 ] ; do
 
     # Cleanup
     if [ ${CLEANUP} -eq 1 ];then
-      echo "$(date -Iseconds) Cleaning up local dir ..."
-      mkdir -p ${DATA_DIR}/geth
-      mv ${DATA_DIR}/geth ${DATA_DIR}/geth.old && rm -rf ${DATA_DIR}/geth.old &
+      echo "$(date -Iseconds) Cleaning up local dir ${GETH_DIR} ..."
+      mkdir -p "${GETH_DIR}"
+      mv "${GETH_DIR}" "${GETH_DIR}.old" && rm -rf "${GETH_DIR}.old" &
       CLEANUP=0
     fi
 
-    # sync from cloud to local disk, without removing existing [missing in the cloud] files
+    # sync from cloud to local disk, with removing existing [missing in the cloud] files
     # run multiple syncs in background
 
-    # we don't wanna sync ancient data here
-    time ${S5CMD} sync ${EXCLUDE_ANCIENT} s3://${STATE_SRC}/* ${CHAINDATA_DIR}/ > cplist_state.txt &
+    time ${S5CMD} sync --delete s3://${STATE_SRC}/* ${STATE_TMP_DIR}/ > cplist_state.txt &
     STATE_CP_PID=$!
-    time nice ${S5CMD} sync --delete --part-size 200 --concurrency 2 ${EXCLUDE_STATE} s3://${ANCIENT_SRC}/* ${CHAINDATA_DIR}/ancient/ > cplist_ancient.txt &
+    time nice ${S5CMD} sync --delete --part-size 200 --concurrency 2 s3://${ANCIENT_SRC}/* ${ANCIENT_TMP_DIR}/ > cplist_ancient.txt &
     ANCIENT_CP_PID=$!
 
     # wait for all syncs to complete
-    # TODO any errors handling here?
-    wait ${STATE_CP_PID} ${ANCIENT_CP_PID}
-
+    # shell tracks all sub-processes and stores exit codes internally
+    # it's not required to stay in wait state for all background processes at the same time
+    # we'll handle these processes sequentially
+    wait ${STATE_CP_PID}
+    STATE_CP_EXIT_CODE=$?
+    wait ${ANCIENT_CP_PID}
+    ANCIENT_CP_EXIT_CODE=$?
+
+    # let's handle out of disk space specially, thus we don't re-try, just stuck here if disk usage is high
+    VOLUME_USAGE_PERCENT=$(df "${DATA_DIR}" | tail -n 1 | awk '{print $5}'|tr -d %)
```

Comment:
WDYT on running volume usage calculations in background, which will periodically (i.e. once per 10 seconds) check the usage and kills sync processes if we are out of space?

Reply:
I think we'll implement it in case of increased demand.  "out-of-space" handling is a simple feature to get out of endless billable loop of `s5cmd` retries.